app/services/compose.py

```diff
@@ -0,0 +1,79 @@
+import inspect
+from typing import Union
+
+from .store.base import BaseStore
+from .messager.base import BaseMessager
+from .inventory import INVENTORY
+
+
+class UnknownImplementationError(Exception):
+    """
+    Raised where we are specifying an unknown implementation.
+    """
+
+    def __init__(self, label: str, service: str):
+        self.msg = f'Implementation "{label}" not found for service: "{service}"'
+
+
+class Composer:
+    """
+    For selecting and (where necessary) instantiating type checked service implementations
+    """
+
+    def __init__(self, config, enforce_base_classes: bool):
+        self.inventory = INVENTORY
+        self.config = config
+        self.enforce_base_classes = enforce_base_classes
+
+    def _get_from_label(self, label: str, implementation_name: str) -> (object):
+        """
+        Select a known implementation based on the str label passed in
+        """
+        implementation = self.inventory[implementation_name].get(label, None)
+        if not implementation:
+            raise UnknownImplementationError(label, implementation_name)
+        return implementation
+
+    def _use(self, implementation: object, base_class: object) -> (object):
+        """
+        Instantiates (where needed) and validates a given impementation complies with
+        the expected Base class before returning it.
+        """
+
+        # Instanitate and configure
+        # note: this is the expected behaviour, we're accounting for pre-instanitation
+        # to allow flexibility for test scenarios.
+        if inspect.isclass(implementation):
+            implementation = implementation()
+            implementation.setup(config=self.config)
+
+        # Confirm implementation is extending the correct base class
+        if self.enforce_base_classes:
+            implementation_in_use_base = implementation.__class__.__bases__[-1]
+            if str(implementation_in_use_base) != str(base_class):
+                raise TypeError(
+                    f"'{implementation_in_use_base}' does not but should have base class '{base_class}'."
+                )
+
+        return implementation
+
+
+    # -----------------------
+    # From here is wrappers, it would be best to continue the pattern as
+    # servies are added
+    
+    def store(self, implementation: Union[str, object]) -> (BaseStore):
+        """
+        Applies sanity checks and setup functions then return the requested store implementation.
+        """
+        if isinstance(implementation, str):
+            implementation = self._get_from_label(implementation, "store")
+        return self._use(implementation, BaseStore)
+
+    def messager(self, implementation: Union[str, object]) -> (BaseMessager):
```

Comment:
The following code may be a slightly nicer way to implement this where the `container` (or `composer` in your implementation) doesn't need to know in advance anything about the types which are registered with it:

```python
from typing import Type, TypeVar, Dict
from abc import ABC, abstractmethod

class Container:

    implementation_map: Dict[Type, Type] = {}

    T = TypeVar("T")

    def register(self, abstract_type: Type[T], implementation_type: Type[T]) -> None:
        self.implementation_map[abstract_type] = implementation_type

    def get(self, abstract_type: Type[T]) -> T:
        """
            Returns an implementation of the abstract_type (or not).
        """
        # todo: error handling
        return self.implementation_map[abstract_type]()


class SomeBaseClass(ABC):
    @abstractmethod
    def something_to_implement(self) -> str:
        ...


class SomeImplementation(SomeBaseClass):
    def something_to_implement(self) -> str:
        return "Hello, world."

ioc_container = Container()
ioc_container.register(SomeBaseClass, SomeImplementation)

some_base_class_imeplementation = ioc_container.get(SomeBaseClass)
print(some_base_class_imeplementation.something_to_implement())
```

The above prints the following btw:

```
Hello, world.
```